File: algorithms/1_easy/D/01089_duplicate_zeros/solution.py
class Solution:
    def duplicateZeros(self, arr: List[int]) -> None:
        arr_len = len(arr)
        count_zeros = 0

        for num in arr:
            if num == 0:
                count_zeros += 1

        new_index = arr_len - 1 + count_zeros
        for i in range(arr_len - 1, -1, -1):
            if new_index < arr_len:
                arr[new_index] = arr[i]

            if arr[i] == 0:
                new_index -= 1
                if new_index < arr_len:
                    arr[new_index] = 0

            new_index -= 1

File: algorithms/1_easy/D/01089_duplicate_zeros/solution.rs
impl Solution {
    pub fn duplicate_zeros(arr: &mut Vec<i32>) {

        let len = arr.len();
        let mut count_zeros = 0;

        for &num in arr.iter() {
            if num == 0 {
                count_zeros += 1;
            }
        }

        let mut new_index = len - 1 + count_zeros;
        for i in (0..len).rev() {
            if new_index < len {
                arr[new_index] = arr[i];
            }

            if arr[i] == 0 {
                new_index -= 1;
                if new_index < len {
                    arr[new_index] = 0; 
                }
            }

            new_index -= 1;
        }
    }
}
